### ISSUE TITLE
wsl --unregister <Distro> should warn users that their data will be permanently deleted.

### ISSUE THREAD

**Issue by jgarzagu** [microsoft/WSL#9932]
### Windows Version

Microsoft Windows [Version 10.0.22621.1413]

### WSL Version

1.1.6.0

### Are you using WSL 1 or WSL 2?

- [X] WSL 2
- [ ] WSL 1

### Kernel Version

5.15.90.1

### Distro Version

Ubuntu 20.04

### Repro Steps

wsl --unregister <Distro>

### Expected Behavior

After running the command line, the program should ask the user to answer yes or no regarding the permanent removal of the distribution.

Example:
```
wsl --unregister <Distro>
Are you sure you want to permanently delete <Distro>? (yes/no)
```

This will prevent more users from mistakenly deleting all their information as it has been reported on multiple occasions. 

For automation purposes, an extra flag should be added to skip the message (e.g., --force).

Example:
```
wsl --unregister <Distro> --force
```

### Actual Behavior

No warnings are prompt to the user. WSL 2 deletes the distribution without asking the user.

**Comment by aki-k** [comment]
Isn't this description enough to tell you what it does?

    --unregister <Distro>
        Unregisters the distribution and **deletes the root filesystem**.

**Comment by jgarzagu** [comment]
@aki-k In my opinion, I think is not enough from the human-computer interaction perspective. 

The commands should be easy to understand even if you don't read the help options for the commands. That's what separates great products from average products. Why do people prefer the Linux terminal over PowerShell in the first place? It is the interaction design, the user experience. The user experience is not just for interfaces. Linux commands are intuitive, and you don't even need to read what all the commands do to understand their functionality.

`wsl --unregister` not only fails to provide a good user experience, but it's becoming more dangerous as WLS is becoming more adopted. First, unregister is something that users doesn't associate with delete. A better name perhaps would be `wsl --delete`. Second, I've seen more tutorials popping up to fix problems that use `wls --unregister` and many unexperienced users who are following the tutorials have reported their distros and all data deleted by accident. 

An example is issue #8525, I also followed those instructions at 1 am and accidentally deleted my distro. It was my mistake, yes, but I never expected a command that would delete everything without even asking, and I was unable to associate unregister with delete. My Ph.D. dissertation was there... 

Please, don't get me wrong, I really like the WSL tools and what the developers are creating, and I want the tools to improve. 🙂

**Comment by aki-k** [comment]
@jgarzagu Yes, you're right and sorry for your loss but the fact still applies that you should understand the commands you run. I actually ran into a Powershell command that didn't ask for confirmation event though I used -Confim:$true and just formatted a filesystem.

**Comment by NotTheDr01ds** [comment]
I think this is a great idea.  I'd even go so far as to say that the prompt should require you to type the name of the distribution to confirm.  I spin up and `--unregister` distributions quite frequently, and I'm still paranoid I'm going to delete the wrong one accidentally.

I like the idea of `--force` for automation.

Only problem, of course, is that the change would break existing applications that automate `--unregister`.

**Comment by NotTheDr01ds** [comment]
As an aside, the other thing I'd love to see related to preventing data-loss is the ability to separately mount additional VHDX's automatically through `/etc/wsl.conf` (per distribution) and `.wslconfig` (for all distributions).

The mount needs to be available *before* the shell is available so that we can mount home directories separate from the distribution itself.  Or (for shared mounts) have a directory of shared scripts that might be called from something like `~/.bashrc`.

Having my home files and important scripts "safe" in another filesystem mount would go a long way towards relieving my paranoia about `--unregister`ing :-).

**Comment by NotTheDr01ds** [comment]
I believe in another thread on the topic, @OneBlue mention the same compatibility issues I did above with automation.  Some thoughts on that:

* Easiest and most compatible:  Perhaps add a warning message, with a delay of 10 seconds (or more) before the distribution is removed.  During that time, the user can Ctrl+C (or another key) to abort, but it will continue automatically if not.  This would allow existing automations to proceed unmodified.

* Add a `--force` option but provide a deprecation period for the old functionality (perhaps in combination with the message/delay above).  After the deprecation period, switch to asking for confirmation unless `--force` is used.

* Have `wsl --unregister` default to moving the `ext4.vhdx` to the Windows Recycle bin or `<userprofile>\temp`.  Remove it with `wsl --unregister --delete` (or one-step with `wsl --unregister --force`).

**Comment by 3kh0** [comment]
That should be a given that the data would be deleted...

**Comment by GiggleMonster** [comment]
So there's no way to get my kali linux distro data back after unregistering?

**Comment by 3kh0** [comment]
Yes, it would have been deleted.

**Comment by ryuheechul** [comment]
I'm actually a bit shocked after finding out how `--unregiser` works...

`--shutdown` works without confirmation and I'm fine with that because the consequences for this is smaller, usually just need to start the distro again and that's it.

But the consequences for `--unregistar` is much greater so it should be not be trivial to do this irreversible action and some UX additions somethings like below should be added:
- requires more explicit options like `--yes` (similar suggested at the top)
  - (or `--force` or user password)
- interactive confirmation when no `--yes` flag is provided

Another approach for improving this workflow would be make this deletion a soft delete not a hard delete so that it's possible to undo it.

So why the safety mechanism for `--shutdown` and `--unregistar` is the same? I think it shouldn't be.

I think this is an UX issue that is bigger threat for people who kind of treats WSL as their "main brain" and using Windows for just the shell. If WSL is promoting the idea of not needing to install Linux into bare metal for personal computing, this one should be improved.

**Comment by jacobus** [comment]
This definitely needs to be fixed. Ideas mentioned above for adding a `--force` parameter, and having a further confirmation is sorely needed. The term (unregister) makes it sound like that you can _unregister_ and _register_ distributions as needed. I was in a hurry and wanted to swop out a troublesome distribution with another and also used `--unregister` with unexpected and bad consequences.  Windows users are accustomed that dangerous commands provide confirmation before running. One should always take an approach of least surprise when developing user interfaces.

**Comment by megapro17** [comment]
Wow, that's a really terrible user experince. It is even possible to unregister without deleting image?

**Comment by yilmazdurmaz** [comment]
"register"ing something has the meaning of acknowledging the existence of something and act accordingly. Thus "unregister" should just remove only the settings needed to run a distro (the acknowledgement) and do not remove its existence for any reason, but instead give details about what to do.

I install a distribution, such as Ubuntu-22.04, only to export it as a template, and then import the same file as many times as I need for different purposes. I remove them when their purpose is met, yet there are times like system crashes that needs a reinstalling the whole system. I had this recently and I can now easily import my previous copies. Saved a lot of time.

But if I accidentally forgot this current "unregister" behavior, and do some operations on the disk until I remember the fatality, it won't be possible to recover them anymore. Thus, my vote goes to this post's purpose: don't delete disks without asking first.

**Comment by GunMetalBull305** [comment]
We are starting to use WSL for development. Having our devs accidently unregister a distro, could be catastrophic. The "unregister" option should simply remove the distro from the registry, and advise the user to manually delete the files.

That usage pattern would bring WSL more inline with other hypervisors, where "removing" a virtual machine only removes its registration, but leave the files intact.

**Comment by MrStLouis** [comment]
I'm in the exact same boat. I just moved a hardrive from an old PC and wanted to get a few files from an old wsl install. I used the --import-in-place on the 4 different distros I somehow had to find the one that I needed. Now I have 4  distros added and I want to clean them up BUT keep the vhdx files in case I need something else at some point. I dont want to unregister because I dont want to delete them, I just dont care to see them as "installed" right now

**Comment by AntonOfTheWoods** [comment]
If you have plenty of spare space you can obviously just export then reimport as often as needed but it would definitely be good to be able to "unregister" and have it do just that, rather than completely delete everything.

`unregister` is patently a horrible verb to use, and clearly doesn't do what most normal speakers of English would expect.

But my guess is that MS never actually expected anyone to do any serious work -> this was just for playing around so you register then unregister your tests when you don't need them anymore. It is almost certain that there is more work going on in WSL than on bare metal Linux installs now, but choices were made back in the day...

**Comment by diegovilar** [comment]
My workaround is to make the virtual drive readonly before unregistering
the distribution. WSL will unregister and silently fail do nuke the file.
Beats having to copy or export the file..

It's still a workaround, of course. "Unregister" is so misleading...

**Comment by GunMetalBull305** [comment]
This is becoming a major detractor for adopting WSL in my developer group. The change required seems straight forward, and would substantially reduce accidental data loss.

If the project contributors will not address this issue, please state so, and close the this ticket. Otherwise, please stop closing duplicate tickets, if you have no intention of addressing this one.

**Comment by aki-k** [comment]
@GunMetalBull305 "major detractor" I never run commands where I don't know what the options do. How can this be so difficult for your group?

**Comment by yilmazdurmaz** [comment]
> @GunMetalBull305 "major detractor" I never run commands where I don't know what the options do. How can this be so difficult for your group?

You might be a very careful person about such issues.
Please, do not think of us lower beings anything close to your standards, and try to feel our pain and have sympathy with your extraordinary existence.

**Comment by aki-k** [comment]
@yilmazdurmaz I agree that it is a problem that --unregister is missing the confirmation, but it's bad practice to run commands without checking out what the options do, like that recursive rm

**Comment by GunMetalBull305** [comment]
@aki-k thanks for agreeing, and for your very constructive and helpful feedback. 👍

**Comment by diegovilar** [comment]
> @yilmazdurmaz I agree that it is a problem that --unregister is missing the confirmation, but it's bad practice to run commands without checking out what the options do, like that recursive rm

I'd say that it's also bad practice to implement a nuke command and name it slap, but maybe that's just me...

The point is that the name of a command, a function, an option, etc. should transmit as much as possible what it will do. Additional actions should come as additional flags, not as a surprise. Nobody expects "unregister" to mean "unregister and delete". Delete, such a destructive and irreversible thing, should be a flag, or at least require confirmation. What it should never be is an unexpected side effect.

Saying that the problem is that the user did not read the documentation is quite frankly outrageous. We name things so we can rapidly understand what they are or do. If not, let's just rename "unregister" to some nonsense word like xqhs72d and I assure you nobody will ever have any expectations about what it does and will refer to the manual. While at it, we could also remove labels from buttons and make the user go to the manual to find out what the third button from the left does...

**Comment by yilmazdurmaz** [comment]
@aki-k there is a reason (almost?) every language has the words to mean "distraction" and "accident". No matter how careful you are, accidents are bound to happen.

But you can, as a programmer, lower the probability of these accidents by naming your functions with conveniently matching functionality. Another way is putting warnings and confirmation prompts to dangerous functions.

Instead of expecting maybe millions of users to be always on the edge when using your program, making a slight change on the program itself to minimize accidents should always be the preferred way.

Our suggestions so far are:
- "--unregister" only unregisters as the name suggest and does not delete the image
  - or moves the image to recycle bin
 - adding another flag, such as "--force" or "--delete", to actually delete the image
 - instead of above ones, adding a confirmation prompt that emphasizes it will delete the image.
 - umm, was there other ways, sorry I forget while writing,

**Comment by GunMetalBull305** [comment]
> @aki-k there is a reason (almost?) every language has the words to mean "distraction" and "accident". No matter how careful you are, accidents are bound to happen.
> 
> But you can, as a programmer, lower the probability of these accidents by naming your functions with conveniently matching functionality. Another way is putting warnings and confirmation prompts to dangerous functions.
> 
> Instead of expecting maybe millions of users to be always on the edge when using your program, making a slight change on the program itself to minimize accidents should always be the preferred way.
> 
> Our suggestions so far are:
> 
> * "--unregister" only unregisters as the name suggest and does not delete the image
>   
>   * or moves the image to recycle bin
> * adding another flag, such as "--force" or "--delete", to actually delete the image
> * instead of above ones, adding a confirmation prompt that emphasizes it will delete the image.
> * umm, was there other ways, sorry I forget while writing,

Those were the options I was thinking of as well. Would be very helpful.

**Comment by aki-k** [comment]
> Microsoft is a large organisation and they change stuff that gets into production very, very rarely. The likelihood of this getting changed is very close to zero, even if they did have devs who cared/agreed.

Here's a $10,000 (maybe more) question: does the following Powershell command confirm from the user whether he wants to format the X: volume?

    PS > Format-Volume -DriveLetter X -FileSystem NTFS -NewFileSystemLabel "NTFS" -Confirm:$True

( Reference: https://learn.microsoft.com/en-us/powershell/module/storage/format-volume )

**Comment by GunMetalBull305** [comment]
> > Microsoft is a large organisation and they change stuff that gets into production very, very rarely. The likelihood of this getting changed is very close to zero, even if they did have devs who cared/agreed.
> 
> Here's a $10,000 (maybe more) question: does the following Powershell command confirm from the user whether he wants to format the X: volume?
> 
> ```
> PS > Format-Volume -DriveLetter X -FileSystem NTFS -NewFileSystemLabel "NTFS" -Confirm:$True
> ```
> 
> ( Reference: https://learn.microsoft.com/en-us/powershell/module/storage/format-volume )

I'm honestly not sure what you are trying to do here. Your comments have contributed nothing, except a little entertainment, to this ticket. We understand and appreciate your viewpoint. There is nothing else you need to say.

**Comment by aki-k** [comment]
@GunMetalBull305 honestly, this ticket shouldn't even have been opened in the first place. and seeing microsoft's non-existing response, this probably won't ever be resolved. happy hunting, captain

**Comment by GunMetalBull305** [comment]
@aki-k thanks again for your wisdom.

**Comment by ohofherr** [comment]
I'm also shocked to find out how `wsl --unregister` works. It should be renamed to `--delete`, `--unregister` should keep the image. A new option `--export-in-place`  that reverses `--import-in-place` wouldn't make much sense from the naming.

**Comment by ChickenIQ** [comment]
Guess who just found this out the hard way while trying to move the data to another drive....
I had backups, so I'm fine, but there command should just unregister the distro, optionally removing the data with something like "--purge"

**Comment by atdiar** [comment]
Yup. Followed blindly some online tips although it corssed my mind that stuff could be deleted.
Scrolled after the fact to see people warning that it would delete everything *smh*

So now, not only is the initial problem not solved (blank black screen with vcxsrv) but I've lost some stuff. Fortunately the most importan (recent code) had been uploaded to github.

Should definitely add some sort of confirmation. -rm -rf silly

**Comment by CarrelV** [comment]
As someone not well versed with WSL/Ubuntu, I tend to rely on ChatGPT to guide me through the different commands to set up a basic env.

When wanting to move a WSL distro from a disk onto a newly installed disk, it does recommend to unregister the distro, then move the leftover .vhdx file to the new disk, and reregister it. It is even adamant about the loss not being expected: [example](https://chatgpt.com/share/675349d6-a3b8-8002-8686-c99d82097306)

Obviously it is my mistake following commands I don't understand without a full backup first, but that is why I was using it in the first place, I am a WSL beginner and was not expecting to erase my distro from existence just like that, without any warning nor clear "delete" name, just "unregister"

**Comment by aki-k** [comment]
What an amazing learning experience, I'll say

**Comment by nihiluis** [comment]
I just had this command last in my bash history (for some reason VScode didn't write to it).

Guess what happened...

**Comment by aki-k** [comment]
@nihiluis It removed your WSL 2 instance like it was supposed to do?

**Comment by iamdevlinph** [comment]
I also had the impression that `--unregister` will just remove the distro from the list or whatever. When on an Administrator powershell, there is also no prompt when running `--unregister` it just completes right away.

**Comment by eric2788** [comment]
What is the command name? `--unregister`

What does it actually do? It unregisters the distribution **AND DELETES YOUR ENTIRE VIRTUAL DISK**.

Come on, something like `--erase` is more fitting than `--unregister`, isn't it?

**Comment by SrinivasTrakMate** [comment]
Is there any tool to recover the deleted distro using wsl --unregister? Anyone had any luck? I had the vhdx in a separate partition and I have not touched it after I deleted the file.

**Comment by yilmazdurmaz** [comment]
> Is there any tool to recover the deleted distro using wsl --unregister? Anyone had any luck? I had the vhdx in a separate partition and I have not touched it after I deleted the file.

If WSL is the only program that writes to that disk, it will be unthouched on the disk. You can use file recovery programs, there are many out there. Some may find directly with folder structure while others will just find file names. Otherwise, there is the risk that fargments might be overwritten. you may still recover partially if these program may find a suitable file table for it.

By the way, I don't know if WSL directly unlinks the file on disk, or moves the file to trashcan. In the latter case, it might be easier to recover. Anyone has an idea/knowledge about that?

**Comment by SrinivasTrakMate** [comment]
Thanks for the reply. I have tried a few things, recuva, testdisk, photorec. Ran these using  ubuntu running on usb. No luck. Only one tool found the vhdx file. It said it is 105GB but it recovered a file of size around 15GB and it was corrupted.

**Comment by yilmazdurmaz** [comment]
> It said it is 105GB but it recovered a file of size around 15GB and it was corrupted.

It is possible that some other program used the disk and corrupted the file,
OR
the operation for some reason failed, yet the program take that error not into account and move on with the data it recovered, thus it is incomplete rather than corrupt.

If the data is very important for you, and the first case is not what hass happeed, please try as many programs as possible until you get the full file or at least usable parts you may work on. 

I am sorry for we can't help any further in this case as we possibly do not have a data recovery expert among us here. (they would probably contact you so far with a service pricing, of course)

**Comment by working-name** [comment]
Nothing about the word unregister hints at deletion. Make it `eradicate` or `destroy-everything` or `erase`, something that actually matches what the command does.

**Comment by SrinivasTrakMate** [comment]
> > It said it is 105GB but it recovered a file of size around 15GB and it was corrupted.
> 
> It is possible that some other program used the disk and corrupted the file, OR the operation for some reason failed, yet the program take that error not into account and move on with the data it recovered, thus it is incomplete rather than corrupt.
> 
> If the data is very important for you, and the first case is not what hass happeed, please try as many programs as possible until you get the full file or at least usable parts you may work on.
> 
> I am sorry for we can't help any further in this case as we possibly do not have a data recovery expert among us here. (they would probably contact you so far with a service pricing, of course)

I don't know is there anyone from wsl team who can tell what exacly does wsl --unregister do? Does it "delete" as in a windows delete command or does something extra? I just trying to figure out a way to recover the vhdx file

**Comment by jon-zuka** [comment]
There should be a `--delete` flag. Unregister I think of unlink. Therefor the command is very dangerous if you use common sense.

**Comment by SnpM** [comment]
An emerging problem is that LLMs may assume `wsl --unregister` isn't dangerous due to the benign connotation of "unregister". As a result, they may recommend the command to uncareful users and cause them harm.

I just did that 😕

**Comment by micsthepick** [comment]
unregister has the wrong affordance in terms of semantics, and yes, sometimes people make assumptions and run commands without reading the description.

**Comment by owenstake** [comment]
So how to unregister distro without delete ext4.vhdx ?

**Comment by yilmazdurmaz** [comment]
> So how to unregister distro without delete ext4.vhdx ?

Basic steps without commands:
- first stop your distro
- find the path to the file
- backup by moving or copying if not too big, you may also export it
  - if moved, unregister may complain. İf that happens, create a new distro, copy its vhdx file, then try again and remove both.

**Comment by kromacie** [comment]
I agree, when you unregister a WSL distro, there’s no warning that the underlying VHDX (your entire Linux filesystem) will be deleted—no confirmation prompt or “safe mode” flag to preserve your disk. Today I set up a new Windows account and tried installing WSL on a non-system drive, since by default everything lands in the system directory and there’s no path option during install. I ran into a permissions error, so I relaunched Terminal as Administrator and re-ran the install.

Because I’d recently been unregistering distros, I assumed it silently skipped deletion this time (one of a common Windows quirks, especially when it comes to WSL). But the moment I ran wsl --unregister, I realized I’d unregistered the wrong distro (as the terminal was actually accessing administrator data)—and I wiped out months of work: uncommitted projects, early-stage code, docs, everything. Fortunately I occasionally back up via Syncthing, so I only lost part of my data. But without that backup, I would have lost everything.

**Comment by voomdoon** [comment]
The installation of a WSL distribution is done via `wsl --install`, which is intuitive and clear.  
However, the operation to remove a distribution is named `wsl --unregister`, which is inconsistent and misleading.

The term “unregister” sounds reversible or administrative — similar to something like “unmount” — and does **not clearly indicate that all data will be permanently deleted without confirmation**.

This mismatch creates a dangerous usability trap, especially for users unfamiliar with the internal consequences. I only realized after losing data that `--unregister` is essentially a full uninstall + purge.

**I strongly suggest renaming `--unregister` to `--uninstall`, or at minimum adding `--uninstall` as an alias, along with a confirmation prompt** (unless a `--force` flag is used).

Thank you for considering this change to improve clarity and prevent accidental data loss.

**Comment by zbingsong** [comment]
Just deleted my distro by mistake, lost everything...

The term "unregister" is very misleading. It sounds more like "unlink," which simply does not give the impression that everything will get purged. 

Plus, there is no confirmation at all, which is simply unreasonable for such a dangerous action. Think about it. When you install/uninstall packages, Ubuntu would ask you to confirm the changes. When you try to delete your github repo, github would ask you to type in the repo name.

Not having a confirmation or a warning is simply a bad design. Yes the consequences are written in the manual, but the assumption that people would read the manual every time they execute a command is, again, unreasonable, especially for a command that looks benign.

**Comment by anthonyL05** [comment]
I just make the same mistake and lose all file...

The current behavior of wsl --unregister is very misleading and dangerous.

There is no interactive warning before deletion. One mistake, and weeks/months of work are lost permanently.

The only mention of this destructive behavior is hidden in wsl --help, but most users do not re-check the help text every time they run a familiar command.

Also, the name itself ("unregister") strongly suggests that it is safe, since "unregister" usually means "deregister" or "remove an entry" — not "permanently delete all data". This misleads users into thinking their distribution will only be removed from the list, not wiped completely  including its ext4.vhdx file.
.

This is not user-friendly and it does not follow the principle of least surprise.

Suggestions:

Rename the command to something explicit like wsl --delete or wsl --remove.

Or, at minimum, display a confirmation prompt:

WARNING: This will permanently delete the entire distribution and all its data. Continue? (y/N)


Provide an option like --force for automation scripts to skip the prompt.

This small change would prevent many users from losing all their work and documentation by mistake.

**Comment by ixfd64** [comment]
I agree. I wanted to do a clean install and am surprised there was no confirmation.

**Comment by notesbytom** [comment]
This existing destructive behavior is described clearly in the output from `wsl --help`, which says `--unregister <Distro> # Unregisters the distribution and deletes the root filesystem.` 
* This is possibly a **more destructive action than a user might expect from the word "unregister" which might imply removing references (registration) but keeping the underlying files for possible re-registration** by something like `wsl --import-in-place`.

The web documentation uses a more verbose description to imply destruction/deletion by `--unregister` but I find the `wsl --help` description more direct and to the point, the root filesystem will be deleted!
* https://learn.microsoft.com/en-us/windows/wsl/basic-commands#unregister-or-uninstall-a-linux-distribution
* _"Caution: Once unregistered, all data, settings, and software associated with that distribution will be permanently lost. Reinstalling from the store will install a clean copy of the distribution."_

I am in favor of some kind of change that allows a Distro to be Unregistered Without Deleting the Underlying VHDX root filesystem so that a user can then possibly re-use the same VHDX with `wsl --import-in-place DISTRO_NAME ROOT\FS\PATH.VHDX` (like trying to Rename a Distro which is oddly a missing feature in WSL).

**Comment by dom6770** [comment]
Yep, that just happened to me. I was trying to rename and move a .vhdx file, so I used `--unregister` intending to move the file and then use `--import-in-place`.

**This is really awful UX.** The `--unregister` flag gives no indication that it also deletes your data. Sure, the `--help` text mentions it, but nobody checks `--help` for every single command.

As the name clearly states, `--unregister` should only remove the distro from WSL, not wipe the underlying VHDX files.
**The fact that there is not even a confirmation prompt, is mind-boggling.**

**Comment by mcduncan-chen** [comment]
Does 'wsl --unregiester' use some kind of TRIM command? File recovery tool seems to be not able to recover the unintented removal of .vhdx file? It's really terrible, single command erases all precious works...
